File: go/scrolling_bitmap.go
package main

// ScrollingBitmap provides a viewport into a larger bitmap.
type ScrollingBitmap struct {
	bitmap      Bitmap
	W, H        int
	XOffset     int
	YOffset     int
	blankPixels []int
}

func NewScrollingBitmap(bm Bitmap, w, h, xOffset, yOffset int) *ScrollingBitmap {
	blank := make([]int, w*h)
	for i := range blank {
		blank[i] = ColorTransparent
	}
	return &ScrollingBitmap{
		bitmap:      bm,
		W:           w,
		H:           h,
		XOffset:     xOffset,
		YOffset:     yOffset,
		blankPixels: blank,
	}
}

func (sb *ScrollingBitmap) Width() int  { return sb.W }
func (sb *ScrollingBitmap) Height() int { return sb.H }

func (sb *ScrollingBitmap) Pixels() []int {
	origW := sb.bitmap.Width()
	origH := sb.bitmap.Height()
	origPixels := sb.bitmap.Pixels()
	pixels := make([]int, len(sb.blankPixels))
	copy(pixels, sb.blankPixels)

	maxX := sb.W
	if origW-sb.XOffset < maxX {
		maxX = origW - sb.XOffset
	}
	maxY := sb.H
	if origH-sb.YOffset < maxY {
		maxY = origH - sb.YOffset
	}
	for y := 0; y < maxY; y++ {
		oy := sb.YOffset + y
		if oy < 0 {
			continue
		}
		for x := 0; x < maxX; x++ {
			ox := sb.XOffset + x
			if ox < 0 {
				continue
			}
			pixels[y*sb.W+x] = origPixels[oy*origW+ox]
		}
	}
	return pixels
}

func (sb *ScrollingBitmap) ScrollTo(x, y int) {
	sb.XOffset = x
	sb.YOffset = y
}

File: ts/src/bitmap/scrolling-bitmap.ts
import type { Bitmap } from "./bitmap.js";
import { Color } from "../server/color.js";

export class ScrollingBitmap implements Bitmap {
  private xOffset: number;
  private yOffset: number;
  private readonly blankPixels: number[];

  constructor(
    private readonly bitmap: Bitmap,
    private readonly width: number,
    private readonly height: number,
    xOffset = 0,
    yOffset = 0,
  ) {
    this.xOffset = xOffset;
    this.yOffset = yOffset;
    this.blankPixels = new Array(width * height).fill(Color.TRANSPARENT);
  }

  getOriginalBitmap(): Bitmap {
    return this.bitmap;
  }

  getWidth(): number {
    return this.width;
  }

  getHeight(): number {
    return this.height;
  }

  getPixels(): number[] {
    const origw = this.bitmap.getWidth();
    const origh = this.bitmap.getHeight();
    const origpixels = this.bitmap.getPixels();
    const pixels = [...this.blankPixels];
    const maxx = Math.min(this.width, origw - this.xOffset);
    const maxy = Math.min(this.height, origh - this.yOffset);
    for (let y = 0; y < maxy; y++) {
      const oy = this.yOffset + y;
      if (oy < 0) continue;
      for (let x = 0; x < maxx; x++) {
        const ox = this.xOffset + x;
        if (ox < 0) continue;
        pixels[y * this.width + x] = origpixels[oy * origw + ox];
      }
    }
    return pixels;
  }

  scrollTo(x: number, y: number): void {
    this.xOffset = x;
    this.yOffset = y;
  }
}
